pkg/build/registry/rest.go

```diff
@@ -24,35 +25,64 @@ var (
 func WaitForRunningBuild(buildClient buildtypedclient.BuildsGetter, build *buildapi.Build, timeout time.Duration) (*buildapi.Build, bool, error) {
 	fieldSelector := fields.OneTermEqualSelector("metadata.name", build.Name)
 	options := metav1.ListOptions{FieldSelector: fieldSelector.String(), ResourceVersion: build.ResourceVersion}
-	w, err := buildClient.Builds(build.Namespace).Watch(options)
-	if err != nil {
-		return build, false, err
-	}
 
-	observed := build
-	_, err = watch.Until(timeout, w, func(event watch.Event) (bool, error) {
-		obj, ok := event.Object.(*buildapi.Build)
-		if !ok {
-			return false, fmt.Errorf("received unknown object while watching for builds: %T", event.Object)
-		}
-		observed = obj
+	done := make(chan interface{}, 1)
+	var resultBuild *buildapi.Build
+	var success bool
+	var resultErr error
 
-		if event.Type == watch.Deleted {
-			return false, ErrBuildDeleted
-		}
-		switch obj.Status.Phase {
-		case buildapi.BuildPhaseRunning, buildapi.BuildPhaseComplete, buildapi.BuildPhaseFailed, buildapi.BuildPhaseError, buildapi.BuildPhaseCancelled:
-			return true, nil
-		case buildapi.BuildPhaseNew, buildapi.BuildPhasePending:
-		default:
-			return false, ErrUnknownBuildPhase
+	deadline := time.Now().Add(timeout)
+	go func() {
+		for time.Now().Before(deadline) {
+			w, err := buildClient.Builds(build.Namespace).Watch(options)
+			if err != nil {
+				resultErr = err
+				done <- true
+				return
+			}
+
+			observed := build
+			_, err = watch.Until(timeout, w, func(event watch.Event) (bool, error) {
+				if event.Type == watch.Error {
+					return false, fmt.Errorf("watch event type error: %v", event)
+				}
+				obj, ok := event.Object.(*buildapi.Build)
+				if !ok {
+					return false, fmt.Errorf("received unknown object while watching for builds: %T", event.Object)
+				}
+				observed = obj
+
+				if event.Type == watch.Deleted {
+					return false, ErrBuildDeleted
+				}
+				switch obj.Status.Phase {
+				case buildapi.BuildPhaseRunning, buildapi.BuildPhaseComplete, buildapi.BuildPhaseFailed, buildapi.BuildPhaseError, buildapi.BuildPhaseCancelled:
+					return true, nil
+				case buildapi.BuildPhaseNew, buildapi.BuildPhasePending:
+				default:
+					return false, ErrUnknownBuildPhase
+				}
+
+				return false, nil
+			})
+
+			if err != nil {
+				resultErr = err
+				done <- true
```

Comment:
hm, i still need to fix this. should not be considered done here if it's a watch error.

Reply:
fixed-ish.